### ISSUE TITLE
Juno Connect with JupyterLab >4.0

### ISSUE THREAD

**Issue by brettc** [rationalmatter/Juno-Issue-Tracker#365]
**Describe the bug**
404 using recent version of jupyterlab

**To Reproduce**
1. Install jupyterlab > 4.0 (on a mac here)
2. run 'jupyterlab lab --ip 0.0.0.0 --port 8888 --no-browser'
3. Check that I can access from another computer on same network (yes / works fine)
4. Open Juno connect (on same network). Attempt to attach
5. Connection light is green / password is requested then 404

**Environment (please complete the following information):**
 - Device: iPad Pro 11
 - iOS version: 16.6
 - App version 1.6.13 (6738)
 
![image](https://github.com/rationalmatter/Juno-Issue-Tracker/assets/457283/8fa09798-4916-4996-8fd8-9baf6296764d)

**Comment by alexstaravoitau** [comment]
Thanks for reporting @brettc — yes, looks like JupyterLab 4.0 now uses Jupyter Notebook 7.0 as its "classic" notebook interface, which is built using new architecture incompatible with Juno Connect. There is no easy fix for this I'm afraid, and in order to support JupyterLab 4.0 I will have to largely rebuild the notebook editor from scratch. I'm currently looking into the best way to do it, but addressing this will take some time I'm afraid — I suggest that you use JupyterLab 3.x with Juno Connect for the time being. Apologies for inconvenience meanwhile... 😔

**Comment by brettc** [comment]
Thanks for the speedy response. Sorry that it is such a big job -- what a pain.

**Comment by alexstaravoitau** [comment]
@brettc Looks like instead of downgrading JupyterLab from 4.0 to 3.x, you could just install the "classic" Notebook as an alternative UI, which will work seamlessly alongside JupyterLab 4.0 installation (and will reuse its backend components):

1. Install "classic" Notebook web app with `pip install nbclassic`.
2. Update the server's URL in Juno Connect to `{SERVER_ADDRESS}/nbclassic/tree`.

Meanwhile, I will still be looking into supporting JupyterLab 4.0 in a more seamless manner.

**Comment by brettc** [comment]
Hey @alexstaravoitau -- thanks for that. I can confirm that this works fine.

**Comment by quoimec** [comment]
Hey @alexstaravoitau - I've been dealing with this problem for the last day or two & have only just found this GH issue. Can I suggest two things:

- You add this as an item in the FAQ for Juno Connect
- And/or intercept the 404 on `tree` requests in app and notify users of the problem

It's totally understandable that this is a difficult issue for you to fix but it is currently way too hard for users to resolve for themselves.

**Comment by alexstaravoitau** [comment]
@quoimec Yes good point, will definitely mention this in the FAQ! Intercepting in the app may be a bit more tricky, but I'll see what I can do.

**Comment by quoimec** [comment]
It seems like if you hit the endpoint host:port/api you'll get the server version back in JSON. I've just tried on my own server and it seems to return this without needing a token/password (slightly concerning). If this is the case, maybe it could be added to the server edit form as part of the reachability test. 

https://jupyter-server.readthedocs.io/en/latest/developers/rest-api.html#get--api-

**Comment by alexstaravoitau** [comment]
@quoimec I've updated the FAQ, so hopefully this workaround will be easier to find: https://juno.sh/troubleshooting-juno-connect/#connecting-to-jupyterlab-4
